### PR TITLE
Add net.toIpAddress, net.toInt

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Add `net.toIpAddress` and `net.toInt` to convert from an `IpAddress` to an integer, and back.
- Inspired (but not copied) from Python 3.10+ `ipaddress` module, simple code, tiny diff.
- Allows to store IP Addresses as 1 primitive integer instead of strings or seqs, because strings sux.
- Documentation, changelog, `since`, `runnableExamples` with `assert` and `static`, bidirectional examples.

---

Python example:

```python
>>> import ipaddress
>>> ipaddress.ip_address(0)
IPv4Address('0.0.0.0')

>>> ipaddress.ip_address(4_294_967_295)
IPv4Address('255.255.255.255')

>>> ipaddress.ip_address(2_130_706_433)
IPv4Address('127.0.0.1')

>>> ipaddress.ip_address(3_232_235_521)
IPv4Address('192.168.0.1')

>>> # With this PR Nim has the same features.
```

🙂

### CODE REVIEW

File: lib/pure/net.nim [hunk 1]
```diff
@@ -2054,3 +2054,29 @@ proc getPrimaryIPAddr*(dest = parseIpAddress("8.8.8.8")): IpAddress =
     result = socket.getLocalAddr()[0].parseIpAddress()
   finally:
     socket.close()
+
+func toIpAddress*(ip: int64): IpAddress {.since: (1, 5).} =
+  ## Convert an `int64` to an `IpAddress` of family `IPv4`.
```
Comment (ringabout) [inline lib/pure/net.nim]:
```suggestion
  ## Converts an `int64` to an `IpAddress` of family `IPv4`.
```

File: lib/pure/net.nim [hunk 2]
```diff
@@ -2054,3 +2054,29 @@ proc getPrimaryIPAddr*(dest = parseIpAddress("8.8.8.8")): IpAddress =
...
+  assert ip >= 0, "IP address must be a positive integer"
+  result = IpAddress(family: IpAddressFamily.IPv4, address_v4: [
+    uint8(ip div 16_777_216'i64 mod 256'i64),
+    uint8(ip div 65_536'i64 mod 256'i64),
+    uint8(ip div 256'i64 mod 256'i64),
+    uint8(ip mod 256'i64),
+  ])
+
+func toInt*(ip: IpAddress): int64 {.since: (1, 5).} =
+  ## Convert an `IpAddress` of family `IPv4` to an `int64`.
```
Comment (ringabout) [inline lib/pure/net.nim]:
```suggestion
  ## Converts an `IpAddress` of family `IPv4` to an `int64`.
```

File: lib/pure/net.nim [hunk 3]
```diff
@@ -2054,3 +2054,29 @@ proc getPrimaryIPAddr*(dest = parseIpAddress("8.8.8.8")): IpAddress =
...
     socket.close()
+
+func toIpAddress*(ip: int64): IpAddress {.since: (1, 5).} =
+  ## Convert an `int64` to an `IpAddress` of family `IPv4`.
+  assert ip >= 0, "IP address must be a positive integer"
+  result = IpAddress(family: IpAddressFamily.IPv4, address_v4: [
+    uint8(ip div 16_777_216'i64 mod 256'i64),
+    uint8(ip div 65_536'i64 mod 256'i64),
+    uint8(ip div 256'i64 mod 256'i64),
+    uint8(ip mod 256'i64),
```
Comment (ringabout) [inline lib/pure/net.nim]: use let ip = `cast[uint64](ip) ` and `shr` stuffs?

File: lib/pure/net.nim [hunk 4]
```diff
@@ -2054,3 +2054,29 @@ proc getPrimaryIPAddr*(dest = parseIpAddress("8.8.8.8")): IpAddress =
...
+      assert 4_294_967_295.toIpAddress.toInt == 4_294_967_295.int64 # 255.255.255.255
+      assert 2_130_706_433.toIpAddress.toInt == 2_130_706_433.int64 # 127.0.0.1
+      assert 3_232_235_521.toIpAddress.toInt == 3_232_235_521.int64 # 192.168.0.1
+
+    example()
+    static: example()
+
+  assert ip.family == IpAddressFamily.IPv4, "IP address must be IPv4"
+  result = int64((16_777_216 * ip.address_v4[0]) +
+    (65_536 * ip.address_v4[1]) + (256 * ip.address_v4[2]) + ip.address_v4[3])
```
Comment (ringabout) [inline lib/pure/net.nim]: cast number to uint64 + `shl`?

File: lib/pure/net.nim (hunk 3 above)
Comment (Araq) [reply lib/pure/net.nim]: Yeah, use bitmasks and shifts here.

File: lib/pure/net.nim (hunk 4 above)
Comment (Araq) [reply lib/pure/net.nim]: Exactly.

**Review by dom96 — COMMENTED** [review]
hmmm, I'm not sure this belongs in the stdlib, or at least in the `net` module.  Also it seems quite specialised and relatively simple to implement manually when needed.

**Comment by Araq** [comment]
Sorry, but effectively nobody asked for these procs and more IPv4 specific stuff seems short-sighted.

**Pull request closed without merging**